### ISSUE TITLE
Offer an option for maximum items to cache

### ISSUE THREAD

**Issue by TimothyGu** [mde/ejs#32]
Implementation open to discussion. LRU? Also in my experiences you need a rwlock for cache reads/writes even though node is single-threaded. For some reason.

**Comment by mde** [comment]
I've never seen any sort of issues with a simple in-process cache that required a R/W lock. In what sort of situation was this occurring? Is this a known issue with Node/IO?

**Comment by TimothyGu** [comment]
I'm not sure. I just added a lock and everything worked fine on TimothyGu/fateserver-node so I didn't investigate further. I'll try to make a test case if I find time.

**Comment by TimothyGu** [comment]
I used https://npmjs.org/package/lru-cache BTW.

**Comment by TimothyGu** [comment]
But for EJS it's probably moot anyway as we do everything in a synchronized fashion.

**Comment by mde** [comment]
I was hoping to avoid adding a lot of code to the client download. What is the tradeoff here?

**Comment by TimothyGu** [comment]
@mde Three options here:

1. NIH an LRU impl. Won't be small either.
2. Find a way to make Browserify skip lru-cache and offer fallbacks.
3. Keep using lru-cache, which isn't *huge* but still makes the minified source bigger by ⅓.

**Comment by mde** [comment]
Option 2 seems the best. More featureful caching seems a lot more useful on the server. This is of course a tradeoff, but I lean very strongly toward keeping EJS small and focused.

The easiest thing might be a small adapter which allows people to override the builtin with the LRU. That doesn't seem to add a huge burden for people who want more sophisticated control, but doesn't require everyone to eat the increased complexity and size.

**Comment by TimothyGu** [comment]
@mde In my last iteration of the PR option 2 is already implemented. I agree that it would be useful to hand the caching operation entirely to the user but I think providing a sane default LRU cache would be helpful as well.

**Comment by mde** [comment]
We definitely want a sane default -- but that minimal implementation there isn't actually LRU, is it?

In any case, the main thing is that the PR adds a lot of complexity in the lib that isn't core functionality, and then there's more code added to flip it off. I'm not super keen on the idea of having to do stuff like use `awk` to rewrite source code.

How about I take a stab at this on the weekend? I just want to start with a dumb object cache by default, and allow some sort of very simple DI for people who care about it to use something more sophisticated.

**Comment by TimothyGu** [comment]
@mde sure no problem. The awk is there just to make testing possible. Also note that the code itself is not complicated at all. The core is only 11 lines, while the remaining 21 lines are all for browser support.

**Comment by mde** [comment]
Fixed in f74cde89e75f8f3982c3f3f122e4ed8aa50390e4, will push in the next release.

A few things:
- Requires a very simple but explicit opt-in (`ejs.cache = LRU(someLimit);`
- There's not a nice API for resetting the LRU cache -- you just blow away ejs.cache with a new instance with your new limit. It looked like this assumed cache was being cleared anyhow, so this seems like a reasonable compromise.
- I rewrote the cache tests to remove the 1-/2-pass approach. The salient test for caching is the final assert in the second pass, and the first pass is mostly just a setup. Ideally tests should be atomic, and having tests with a lot of asserts in them is generally preferable to having tests that depend on side-effects from previous tests.
- I kind of half-assed the JSDoc for the simple cache implementation in utils. Happy to revisit it when there's time.
